### PR TITLE
Attempt to improve codegen for arrays of repeated enums

### LINKED ISSUE #101685: Repeating an enum does not generate a memset
I tried this code: [godbolt](https://rust.godbolt.org/z/WT4v9Mnar)

```rust
pub fn create_array() -> [Option<u8>; 256] {
    [None; 256]
}
```

I expected to see this happen: This should compile to a memset, or the equivalent unrolled SIMD stores

[LLVM IR](https://rust.godbolt.org/z/qdG4hPMEc)

Instead, this happened: It compiles to a bunch of movs.

I think this happens because we transform this to a bunch of stores of the discriminant, without touching the data part of the option, making it impossible for LLVM to memset this. Ideally, we'd write 0 to the data part as well here, allowing this to use memset.

### PR DESCRIPTION
**Pull request by clubby789**
Fixes #101685

For enums where the tag and value are the same type, we can get better codegen for `[Enum; N]` by emitting a vector store instruction.

1024 as a limit was picked randomly and is probably too low - I'm unsure how to construct a vector of alternating elements without it being O(n)

### CODE REVIEW

File: src/test/codegen/enum-repeat.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    // CHECK-NEXT: ret void
+    [None; 64]
+}
+
+// CHECK-LABEL: @some_repeat
+#[no_mangle]
+pub fn some_repeat() -> [Option<u8>; 64] {
+    // CHECK: store <128 x i8>
+    // CHECK-NEXT: ret void
+    [Some(0); 64]
```
Comment (scottmcm) [inline src/test/codegen/enum-repeat.rs:18]: It's interesting to me that `some_repeat` seems to do 16 bytes at a time on x64, but `none_repeat` doesn't.  Might be interesting to look at why LLVM is treating them differently.

Also, out of curiosity, what's the assembly difference before/after this PR?

**Comment by scottmcm** [comment]
> I'm unsure how to construct a vector of alternating elements without it being O(n)

No idea if this is a *good* way, but the way that jumps to mind is to [`shufflevector`](https://llvm.org/docs/LangRef.html#shufflevector-instruction) together shorter ones double the length each time.  Though that's still O(n) in the mask vector, so maybe using memory operations to do it would be better.

That said, the other thing that comes to mind is that it might be easier to lower the `[x; N]` to a call to some lang-item `core::array::repeat::<_, N>(x)`, so these optimizations could be done in Rust code instead of codegen.  Then other backends would get them too, and it's not obvious to me that making 1024-element vectors is the best way here.  Is more than 64 bytes at once ever better, on current machines?  In rust code it could, for example, loop over [`as_chunks_mut`](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.as_chunks_mut) to write chunks at a time and let LLVM then optimize that.

File: compiler/rustc_codegen_llvm/src/builder.rs [hunk 2]
```diff
@@ -561,6 +561,24 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
...
+            v1 = self.from_immediate(v1);
+            v2 = self.from_immediate(v2);
+            let ty = self.cx().val_ty(v1);
+            // Create a vector of size 2*count and store it in one instruction
+            if ty == self.cx().val_ty(v2) {
+                let count = count * 2;
+                let vec = unsafe { llvm::LLVMGetUndef(self.type_vector(ty, count as u64)) };
+                let vec = (0..count as usize).fold(vec, |acc, x| {
+                    let elt = [v1, v2][x % 2];
+                    self.insert_element(acc, elt, self.cx.const_i32(x as i32))
```
Comment (scottmcm) [inline compiler/rustc_codegen_llvm/src/builder.rs:574]: No idea if this is better, but you could do it in O(1) LLVM instructions by making a `<2 x _>` (with two insert_elements) and then repeating that one with a shuffle like
```llvm
shufflevector <2 x i8> %x, <2 x i8> undef, <64 x i32> <i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1, i32 0, i32 1>
```
https://llvm.godbolt.org/z/5feKaEo5z

File: src/test/codegen/enum-repeat.rs (hunk 1 above)
Comment (clubby789) [reply src/test/codegen/enum-repeat.rs:18]: [godbolt](https://godbolt.org/z/68xrb14EG) - This is `none_repeat` before and after LLVM opts - it looks like LLVM is optimising out the undefined write and forgetting about it. This happened even if I explicitly added `store undef`.

The current assembly output is
- `None; 64` - 64 single byte `mov`s to the tag
- `Some(0); 64` - A 16 byte constant vector (alternating 0/1) is stored into the array with `movups`
- `Some(1); 64` - The same as `Some(0)`, but with an all-1's pattern

With this patch
- `None; 64` - 16 byte vector created with `xorps`, then `movups`
- `Some(0); 64`/`Some(1); 64` - Same as before

**Comment by clubby789** [comment]
I've implemented it using `shufflevector` now, but I've noticed that this means that cases like `Some(1); 1024` (which used to be a memset) are now a long series of vector stores. This seems like something LLVM should be handling but isn't - I'll see if I can get better emitted code.

@rustbot author

**Comment by clubby789** [comment]
My original approach worked at the `rustc_codegen_ssa` level and did a `memset` with the tag value - that improved the `None` case, but the `Some(1)` case got much worse. Something along those lines might help

**Comment by clubby789** [comment]
I'm going to close this since it only targets LLVM, is a hacky way of solving the problem, and actually causes regressions 😅

**Pull request closed without merging**